rosbag2_compression/src/rosbag2_compression/sequential_compression_reader.cpp

```diff
@@ -0,0 +1,198 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rosbag2_compression/sequential_compression_reader.hpp"
+
+#include <memory>
+#include <stdexcept>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "rcpputils/filesystem_helper.hpp"
+#include "rosbag2_compression/compression_options.hpp"
+#include "rosbag2_compression/zstd_decompressor.hpp"
+#include "logging.hpp"
+
+
+namespace rosbag2_compression
+{
+
+SequentialCompressionReader::SequentialCompressionReader(
+  std::unique_ptr<rosbag2_storage::StorageFactoryInterface> storage_factory,
+  std::shared_ptr<rosbag2_cpp::SerializationFormatConverterFactoryInterface> converter_factory,
+  std::unique_ptr<rosbag2_storage::MetadataIo> metadata_io)
+: storage_factory_{std::move(storage_factory)},
+  converter_factory_{std::move(converter_factory)},
+  metadata_io_{std::move(metadata_io)}
+{}
+
+SequentialCompressionReader::~SequentialCompressionReader()
+{
+  reset();
+}
+
+void SequentialCompressionReader::reset()
+{
+  storage_.reset();
+}
+
+void SequentialCompressionReader::setup_compression()
+{
+  compression_mode_ = rosbag2_compression::compression_mode_from_string(metadata_.compression_mode);
+  if (compression_mode_ != rosbag2_compression::CompressionMode::NONE) {
+    if (metadata_.compression_format == "zstd") {
+      decompressor_ = std::make_unique<rosbag2_compression::ZstdDecompressor>();
+      // Decompress the first file so that it is readable.
+      ROSBAG2_COMPRESSION_LOG_DEBUG_STREAM("Decompressing " << get_current_file().c_str());
+      set_current_file(decompressor_->decompress_uri(get_current_file()));
+    } else {
+      std::stringstream err;
+      err << "Unsupported compression format " << metadata_.compression_format;
+      throw std::invalid_argument{err.str()};
+    }
+  }
+}
+
+void SequentialCompressionReader::open(
+  const rosbag2_cpp::StorageOptions & storage_options,
+  const rosbag2_cpp::ConverterOptions & converter_options)
+{
+  if (metadata_io_->metadata_file_exists(storage_options.uri)) {
+    metadata_ = metadata_io_->read_metadata(storage_options.uri);
+    if (metadata_.relative_file_paths.empty()) {
+      ROSBAG2_COMPRESSION_LOG_WARN("No file paths were found in metadata.");
+      return;
+    }
+    file_paths_ = metadata_.relative_file_paths;
+    current_file_iterator_ = file_paths_.begin();
+    setup_compression();
+
+    storage_ = storage_factory_->open_read_only(
+      *current_file_iterator_, metadata_.storage_identifier);
+    if (!storage_) {
+      throw std::runtime_error{"No storage could be initialized. Abort"};
+    }
+  } else {
+    throw std::runtime_error{"Compression is not supported for legacy bag files."};
+  }
+  auto topics = metadata_.topics_with_message_count;
```

Comment:
Shouldn't this be a reference?
```suggestion
  auto & topics = metadata_.topics_with_message_count;
```